### ISSUE TITLE
Wrong error message when Self gets a different type parameter in implementation

### ISSUE THREAD

**Issue by uwap** [rust-lang/rust#17700]
If you use Self within a trait and you mix the type parameter up in the implementation it will give you an error message saying `method bar has an incompatible type for trait: expected type parameter, found type parameter`. This sounds a bit weird.  
Some example code:

``` rust
trait Foo<A> {
  fn bar<B>(&self, b: B) -> Self;
}

impl <A> Foo<A> for Option<A> {
  fn bar<B>(&self, b: B) -> Option<B> { // has to be Option<A>, throws "expected type parameter, found type parameter"
    Some(b)
  }
}
```

**Comment by nham** [comment]
Triage: It seems like this is still an issue. The error message is better now:

```
17700.rs:6:3: 8:4 error: method `bar` has an incompatible type for trait: expected type parameter, found a different type parameter [E0053]
17700.rs:6   fn bar<B>(&self, b: B) -> Option<B> {
17700.rs:7     Some(b)
17700.rs:8   }
error: aborting due to previous error
```

But it seems like there is room for improvement, like underlining which parameter is bad.

**Comment by estebank** [comment]
Looks much better now:
```nocode
error[E0053]: method `bar` has an incompatible type for trait
 --> <anon>:6:29
  |
2 |   fn bar<B>(&self, b: B) -> Self;
  |                             ---- type in trait
...
6 |   fn bar<B>(&self, b: B) -> Option<B> {
  |                             ^^^^^^^^^ expected type parameter, found a different type parameter
  |
  = note: expected type `fn(&std::option::Option<A>, B) -> std::option::Option<A>`
  = note:    found type `fn(&std::option::Option<A>, B) -> std::option::Option<B>`
```

**Comment by uwap** [comment]
This looks great. Thanks.
